### PR TITLE
Create tool lug -- lookup generator

### PR DESCRIPTION
**Pull request by gonzus**
Profiling revealed that the most called function is
get_alphabet_position(), which is implemented using a partial lookup
table.  Rather than having this partial version of a lookup table, which
required a bit more runtime logic to choose the correct entry, we simply
generate a full lookup table for all possible characters and index
directly into it.  This full lookup table is then included as part of
the source during compilation, to speed things up at runtime.

See instructions in c/README.md for how to regenerate the lookup table.

### CODE REVIEW

**Comment by gonzus** [comment]
BTW, if this change makes sense to you guys, I will be happy to submit a similar PR for the C++ version.

File: c/olc_lookup.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// THIS FILE WAS GENERATED AUTOMATICALLY
+// DO NOT EDIT IT UNLESS YOU KNOW WHAT YOU ARE DOING
+
+static int kAlphabetPositionLUT[256] = {
```
Comment (zongweil) [inline c/olc_lookup.h]: Can you generate a comment that explains what this is? It's not obvious looking at it what it represents and what each row/column represents.

**Comment by zongweil** [comment]
Can you use your benchmark script to get a sense of the difference in performance?

**Comment by drinckes** [comment]
I'm kind of uncomfortable with rewriting the code in this way (making the constant definitions complicated) without examining how the algorithm is implemented, or e.g., converting it to a class so that you can build this structure at class initialisation time (i.e. once) and then running a million requests through it.

(Like, if you have a billion lat/lngs to convert, and you care about speed, you don't want to do that by running a command line for each one.)

How about raising performance as an issue or on the public mailing list and seeing if this is actually a problem?

File: c/olc_lookup.h (hunk 1 above)
Comment (gonzus) [reply c/olc_lookup.h]: Added with commit 3f786403e5b620f50f5544c7bf6ea0a8d4982f04

**Comment by gonzus** [comment]
> Can you use your benchmark script to get a sense of the difference in performance?

Added with commit 6112426e2e326cd9dd51a00095ec7a8d43e10e0e

File: c/tools/Makefile [hunk 2]
```diff
@@ -0,0 +1,19 @@
+all: lug
+
+C_FLAGS += -Wall
+# C_FLAGS += -Wextra
```
Comment (IDisposable) [inline c/tools/Makefile]: Commented-out code shouldn't be checked in

File: c/Makefile [hunk 3]
```diff
@@ -12,8 +12,9 @@ C_FLAGS += -Wno-comment
 LD_FLAGS += -lm
 
 ALL_FLAGS += -std=c99
-ALL_FLAGS += -g
-ALL_FLAGS += -pg
+# ALL_FLAGS += -g
```
[the remark this answers is not in the document]
Comment (gonzus) [reply c/Makefile]: It is commented out code, you are right. I usually leave those there to quickly change from a debug build to an optimized one, or to turn profiling on. But I cleaned those up anyway with c2b1fc443ab3751c45ddf3e67b46bcf06f24d464.

File: c/Makefile [hunk 4]
```diff
@@ -23,7 +24,7 @@ LIB_OBJ_FILES = $(LIB_C_FILES:.c=.o)
 LIB_A_FILE = lib$(LIB_NAME).a
 
 %.o : %.c
-	$(CC) -c $(ALL_FLAGS) $(C_FLAGS) $(CPP_FLAGS) -o $@ $< $(LD_FLAGS)
+	$(CC) -c $(ALL_FLAGS) $(C_FLAGS) $(CPP_FLAGS) -o $@ $<
```
[the remark this answers is not in the document]
Comment (gonzus) [reply c/Makefile]: Yes it is. `LD_FLAGS` only apply when building an executable binary, not an object file.

File: c/tools/Makefile (hunk 2 above)
Comment (gonzus) [reply c/tools/Makefile]: Resolved with same commit as above, c2b1fc443ab3751c45ddf3e67b46bcf06f24d464.

**Comment by gonzus** [comment]
Hi @zongweil -- sure, no problem. I will close this PR and move the code into a separate branch in my clone. Perhaps I will send another PR in the near future with some minor restructuring of the files, so that it is easier for me to maintain the unrolled LUT -- and maybe I could send (after that) another PR that adds the unrolled LUT as a compile-time option?

I will take a look at diving into the C++ implementation to check my theory about those string functions.

Cheers!

**Pull request closed without merging**